common/src/main/java/com/microsoft/identity/common/internal/telemetry/Telemetry.java

```diff
@@ -0,0 +1,283 @@
+            return;
+        }
+
+        synchronized (this) {
+            //check the configuration
+            if (!mDefaultConfiguration.isDebugEnabled() && mIsDebugging) {
+                return;
+            }
+
+            List<Map<String, String>> finalRawMap = new ArrayList<>();
+
+            for (Map<String, String> event : mTelemetryRawDataMap) {
+                if (correlationId.equalsIgnoreCase(event.get(TELEMETRY_KEY_CORRELATION_ID))) {
+                    finalRawMap.add(applyPiiOiiRule(event));
+                    finalRawMap.remove(event);
+                }
+            }
+
+            //Add the telemetry context to the telemetry data
+            finalRawMap.add(applyPiiOiiRule(mTelemetryContext.getProperties()));
+
+            for (ITelemetryObserver observer : mObservers) {
+                observer.send(finalRawMap);
+            }
+        }
+    }
+
+    private Map<String, String> applyPiiOiiRule(final Map<String, String> properties) {
+        if (mDefaultConfiguration.isPiiEnabled()) {
+            Logger.warn(TAG, "Telemetry PII/OII is enabled by the developer.");
+            return properties;
+        }
+
+        final Map<String, String> nonPiiProperties = new HashMap<>();
+        for (Map.Entry<String, String> entry : properties.entrySet()) {
+            if (!TelemetryPiiOiiRules.getInstance().isPiiOrOii(entry.getKey())) {
+                nonPiiProperties.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        return nonPiiProperties;
+    }
+
+    /**
+     * API for creating {@link Telemetry} instances.
+     */
+    public static class Builder {
+        private Context mContext;
+        private TelemetryConfiguration mDefaultConfiguration;
+        private TelemetryContext mTelemetryContext;
+        private Boolean mIsDebugging;
+
+        public Builder() {
+        }
+
+        public Builder withContext(final Context context) {
+            if (context == null) {
+                throw new IllegalArgumentException("Context must not be null.");
+            }
+            if (!hasPermission(context, Manifest.permission.INTERNET)) {
+                throw new IllegalArgumentException("INTERNET permission is required.");
+            }
+            mContext = context.getApplicationContext();
+            if (mContext == null) {
+                throw new IllegalArgumentException("Application context must not be null.");
+            }
+
+            mTelemetryContext = TelemetryContext.create(mContext);
+
+            try {
+                String packageName = context.getPackageName();
+                int flags = context.getPackageManager().getApplicationInfo(packageName, 0).flags;
+                mIsDebugging = (flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;
+            } catch (final PackageManager.NameNotFoundException exception) {
+                Logger.warn(TAG, "The application is not found from PackageManager.");
+                mIsDebugging = true;
```

Comment:
In which scenario would this happen? Can this happen in prod? Why are we setting debugging = true in this case?